service/api_task.go

```diff
@@ -635,6 +634,26 @@ func handleOldAgentRevision(response apimodels.NextTaskResponse, h *host.Host, w
 			gimlet.WriteJSON(w, response)
 			return apimodels.NextTaskResponse{}, true
 		}
+
+		// Non-legacy hosts deploying agents via the agent monitor may be
+		// running an agent on the current revision, but the database host has
+		// yet to be updated.
+		if !h.LegacyBootstrap() && details.AgentRevision != h.AgentRevision {
```

Comment:
You might also want to check if the host is legacy on line 615. I think setting that will make us deploy the agent in the old manner

Reply:
It shouldn't matter if the `Host.NeedsNewAgent` flag is set or not on non-legacy hosts. The agent deploy job is never supposed to run on non-legacy hosts, so I made the `host.NeedsNewAgentFlagSet()` query (which populates the agent deploy job) filter out non-legacy hosts.